<a href="https://colab.research.google.com/github/ahsan-fayyaz/Artificial-Intelligence/blob/main/Assignment_4/Assignment_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.datasets import load_iris
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

import warnings
warnings.filterwarnings('ignore')

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
iris_data = load_iris()

features = iris_data.data
target =  iris_data.target

In [3]:
# Output the first 15 rows of the data
df_features = pd.DataFrame(features, columns=iris_data.feature_names)
df_target = pd.DataFrame(target, columns=['target'])

df = pd.concat([df_features, df_target], axis=1)
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  
10       0  
11       0  
12       0  
13       0  
14       0

In [4]:
# Display a summary of the table information (number of datapoints, etc.)
display(df.describe())
display(df.info())

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal length (cm)    150 non-null float64
sepal width (cm)     150 non-null float64
petal length (cm)    150 non-null float64
petal width (cm)     150 non-null float64
target               150 non-null int64
dtypes: float64(4), int64(1)
memory usage: 5.9 KB


None

## About the dataset

Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?


In [5]:
print("Feature Names : ",iris_data.feature_names)
print("Target Names  : ",iris_data.target_names)

Feature Names :  ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Target Names  :  ['setosa' 'versicolor' 'virginica']


> After exploring the iris dataset, it can be observed that there are 4 features (4 columns)
> 1. Septal Length (cm)
> 2. Septal Width  (cm)
> 3. Petal Length  (cm)
> 4. Petal Width   (cm)

> The label in the dataset divides into species classes (3 categories) -
> where the values 0,1,2 correspond to the following different species
> 1. Iris setosa (value = 0)
> 2. Iris versicolor (value = 1)
> 3. Iris virginica (value = 2)


# Part 2: Split the dataset into train and test


In [6]:
# Take the dataset and split it into our features (X) and label (y)
X = features
y = target

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=50)


# Part 3: Logistic Regression


In [7]:
# i. Use sklearn to train a LogisticRegression model on the training set
LR_model = LogisticRegression()
LR_model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [8]:
# ii. For a sample datapoint, predict the probabilities for each possible class
LR_sample_dataset = X_test[:1]
LR_pred = LR_model.predict_proba(LR_sample_dataset)
print("Predictions Probability: ", LR_pred)

Predictions Probability:  [[0.05591523 0.65913963 0.28494514]]


In [9]:
# iii. Report on the score for Logistic regression model, what does the score measure?
# Here is the classification report for the model
LR_y_pred = LR_model.predict(X_test)
print(classification_report(y_test, LR_y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         6
           1       1.00      1.00      1.00         4
           2       1.00      1.00      1.00         5

   micro avg       1.00      1.00      1.00        15
   macro avg       1.00      1.00      1.00        15
weighted avg       1.00      1.00      1.00        15



In [10]:
#Here is the confusion matrix
print("Confusion Matrix :\n",confusion_matrix(y_test, LR_y_pred))
print("\n")
print("Score : ", LR_model.score(X_test, y_test))

Confusion Matrix :
 [[6 0 0]
 [0 4 0]
 [0 0 5]]


Score :  1.0


#### Report on the score:
> I ran a classification report on the model which determined precision of the model. The precision is 1.0

> Precision measures how good our model is when the prediction is positive. The score is the coefficient of prediction, representing the model's capability to predict the correct classes. Here, we have a perfect score of 1.0

> I also printed confusion matrix for this model and here are the results:
> * 6 out of 6 True positives for setosa
> * 6 out of 6 True positives for versicolor
> * 3 out of 3 True positives for virginica

In [11]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
print("Coefficients : ", LR_model.coef_)
print("\n")
print("Intercepts: ", LR_model.intercept_)

Coefficients :  [[ 0.37913418  1.44405025 -2.20416393 -0.99641481]
 [ 0.51982378 -1.63674234  0.44256066 -1.22218456]
 [-1.59429958 -1.63794573  2.43430576  2.38856193]]


Intercepts:  [ 0.26022165  0.96602741 -1.1157134 ]



# Part 4: Support Vector Machine


In [12]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVM_model = svm.SVC(kernel="linear", probability=True)
SVM_model.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='linear', max_iter=-1, probability=True, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [13]:
# ii. For a sample datapoint, predict the probabilities for each possible class
SVM_sample_dataset = X_test[:1]
SVM_pred = SVM_model.predict_proba(SVM_sample_dataset)
print("Predictions Probability: ", SVM_pred)

Predictions Probability:  [[0.01319876 0.97526335 0.01153789]]


In [14]:
# iii. Report on the score for the SVM, what does the score measure?

SVM_y_pred = SVM_model.predict(X_test)
print("\nClassification Report : ",classification_report(y_test, SVM_y_pred))


Classification Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00         6
           1       1.00      0.75      0.86         4
           2       0.83      1.00      0.91         5

   micro avg       0.93      0.93      0.93        15
   macro avg       0.94      0.92      0.92        15
weighted avg       0.94      0.93      0.93        15



In [15]:
#Here is the confusion matrix
print("\nConfusion Matrix : ",confusion_matrix(y_test, SVM_y_pred))
print("\n")
SVM_score = SVM_model.score(X_test, y_test)
print("Score : ", SVM_score)


Confusion Matrix :  [[6 0 0]
 [0 3 1]
 [0 0 5]]


Score :  0.9333333333333333


#### Report on the score:
> For SVM, We have a score of 0.933. This depicts that the model fits pretty well although it doesn't have a perfect score like our Logistic Regression. Overall, the model fits pretty well giving us a score of 0.93 in SVM model

#  Part 5: Neural Network


In [16]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
NN_model = MLPClassifier(
            solver='lbfgs', 
            alpha=1e-5, 
            hidden_layer_sizes=(12, 6), 
            random_state=1)
NN_model.fit(X_train, y_train)

MLPClassifier(activation='relu', alpha=1e-05, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(12, 6), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=1, shuffle=True, solver='lbfgs', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [17]:
# ii. For a sample datapoint, predict the probabilities for each possible class
NN_sample_dataset = X_test[:1]
NN_pred = NN_model.predict_proba(NN_sample_dataset)
print("Predictions: ", NN_pred)

Predictions:  [[4.33214753e-09 9.99994040e-01 5.95559621e-06]]


In [18]:
# iii. Report on the score for the Neural Network, what does the score measure?
NN_y_pred = NN_model.predict(X_test)
print("\nClassification Report : ",classification_report(y_test, NN_y_pred))


Classification Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00         6
           1       1.00      0.50      0.67         4
           2       0.71      1.00      0.83         5

   micro avg       0.87      0.87      0.87        15
   macro avg       0.90      0.83      0.83        15
weighted avg       0.90      0.87      0.86        15



In [19]:
#Here is the confusion matrix
print("\nConfusion Matrix : ",confusion_matrix(y_test, NN_y_pred))
print("\n")
NN_score = NN_model.score(X_test, y_test)
print("Score : ", NN_score)


Confusion Matrix :  [[6 0 0]
 [0 2 2]
 [0 0 5]]


Score :  0.8666666666666667


#### Report on the score:
> I used the following parameters here:  
> * solver='lbfgs', 
> * alpha=1e-5, 
> * hidden_layer_sizes=(12, 6), 
> * random_state=1

> The model gave a score of 0.87 here. Our model predicts correct 86% of the times. We see that there were some true negatives in our result as shown in the confusion matrix

In [20]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

###Use sklearn to train a Neural Network (MLP Classifier) on the training set
NN_model_2 = MLPClassifier(
            learning_rate = "adaptive",
            max_iter = 100,
            activation = "identity",
            alpha=1e-5, 
            hidden_layer_sizes=(12, 6), 
            random_state=0)
NN_model_2.fit(X_train, y_train)

###For a sample datapoint, predict the probabilities for each possible class
NN_sample_dataset_2 = X_test[:1]
NN_pred_2 = NN_model_2.predict_proba(NN_sample_dataset_2)
print("Predictions: ", NN_pred_2)

###Report on the score for the Neural Network, what does the score measure?
NN_y_pred_2 = NN_model_2.predict(X_test)
print("\nClassification Report : ",classification_report(y_test, NN_y_pred_2))

###Here is the confusion matrix
print("\nConfusion Matrix : ",confusion_matrix(y_test, NN_y_pred_2))
print("\n")
NN_score_2 = NN_model_2.score(X_test, y_test)
print("Score : ", NN_score_2)

Predictions:  [[0.22360226 0.42756553 0.3488322 ]]

Classification Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00         6
           1       1.00      0.25      0.40         4
           2       0.62      1.00      0.77         5

   micro avg       0.80      0.80      0.80        15
   macro avg       0.88      0.75      0.72        15
weighted avg       0.88      0.80      0.76        15


Confusion Matrix :  [[6 0 0]
 [0 1 3]
 [0 0 5]]


Score :  0.8


#### Report on the score:
> I used the following parameters here on the second version of MLP:
> * learning_rate = "adaptive",
> * max_iter = 100,
> * activation = "identity",
> * alpha=1e-5, 
> * hidden_layer_sizes=(12, 6), 
> * random_state=0

> The model gave a score of 0.80
> It seems as if the model became less accurate when I switched max_iter = "100" and random_state=0. Also, I changed activation parameter = "identity" which might also be the reason.

# Part 6: K-Nearest Neighbors
 

In [21]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
KNN_model = KNeighborsClassifier(n_neighbors=5)
KNN_model.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [22]:
# ii. For a sample datapoint, predict the probabilities for each possible class
KNN_sample_dataset = X_test[:1]
KNN_pred = KNN_model.predict_proba(KNN_sample_dataset)
print("Predictions: ", NN_pred)

Predictions:  [[4.33214753e-09 9.99994040e-01 5.95559621e-06]]


In [23]:
# iii. Report on the score for kNN, what does the score measure?
KNN_y_pred = KNN_model.predict(X_test)
print("\nClassification Report : ",classification_report(y_test, KNN_y_pred))


Classification Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00         6
           1       1.00      0.75      0.86         4
           2       0.83      1.00      0.91         5

   micro avg       0.93      0.93      0.93        15
   macro avg       0.94      0.92      0.92        15
weighted avg       0.94      0.93      0.93        15



In [24]:
#Here is the confusion matrix
print("\nConfusion Matrix : ",confusion_matrix(y_test, KNN_y_pred))
print("\n")
KNN_score = KNN_model.score(X_test, y_test)
print("Score : ", KNN_score)


Confusion Matrix :  [[6 0 0]
 [0 3 1]
 [0 0 5]]


Score :  0.9333333333333333


#### Report on the score:
> We have a score of 0.933 again. This depicts that the model fits pretty well to the data and gives us a fairly accurate model.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?


> I ran 4 different algorithms in this notebook:
> * Logistic Regression
> * Support Vector Machine
> * MLP Classifier (Neural Network)
> * K-Nearest Neighbors

> The models scored an accuracy between 80% and 100%. It is safe to say that the models did fairly well and the outputs did change upon changing the parameters. Logistic Regression was able to give us a 100% accuracy. However, it may also be the reason that we have a pretty small dataset (150 rows) and it may be due to overfitting.

> The observation that surprised me was that max_iter parameter in Neural Network (MLP Classifier) had a huge impact on the score value. The score went really down when I put max_iter=10, however, I changed it to 100 for the second variation of the algorithm and it gave me a score of 80%. Overall, it is safe to say that Logistic regression made most sense due to the fact that it constructs linear boundaries to classify objects. 